### PR TITLE
Replaced global ContextShift/Timer instances on JVM

### LINKED ISSUE #509: Proposal: Don't default to ExecutionContext.global in IOApp
So `IOApp` appears to default to a `ContextShift[IO]` based on `ExecutionContext.global`. This is a semi-sane default on the surface and it aligns well with other Scala applications, but it seems to cause a ton of problems in practice due to the fact that it's backed by a fork-join pool. I've observed absolutely horrible things like over 250 threads sitting in that pool (on an 8 core machine!), spawned by fork-join in response to long-lived CPU-bound tasks.

Fork-Join appears to be optimized with two assumptions in mind. First, it assumes that people are lazy-moding their pool usage and dropping blocking IO in the middle of their compute pool. This assumption triggers pool growth well beyond num CPUs when threads aren't being returned to the pool in a timely manner, even when the *reason* those threads aren't being returned is simply long-lived CPU-bound computations. Second, it seems to assume that people are tuning their applications for latency rather than throughput, and thus frequently returning control back to the thread pool even when CPU-bound. 

Both of these assumptions mesh very well with typical use of `scala.concurrent.Future`, which yields on every `map` and is frequently used for blocking IO (mostly because many people simply don't realize this is bad). However, due to some very intentional and very carefully thought out API design, cats-effect (and all frameworks within its ecosystem) tends to do a very good job of pushing users to keep their blocking IO off the main pool, *and* also tends to bias towards throughput rather than latency (due to explicit rather than implicit yield semantics). This makes fork-join *in general* a very, very poor fit for cats-effect, fundamentally.

And this is all ignoring the perils of the global pool in general and the fact that many frameworks abuse it!

All of this, in general, suggests that we would be better off creating our own fixed thread pool in `IOApp`. We experimented with this recently at SlamData and saw our context shift time massively decrease as soon as we got off of fork-join, and we're not doing anything weird at all in terms of our pool usage (additionally, none of our dependencies are using the global pool at all, as verified in thread dumps, so it was definitely just our own compute-bound pool usage that was triggering this behavior).

Anyway, that's my proposal and the rationale. I just really, really don't think it's a good idea for `IOApp` to be defaulting to fork-join *anything*, since I don't believe that any app written in the cats-effect ecosystem is going to behave particularly well on fork-join for fundamental reasons.

### LINKED ISSUE #515: IOApp default contextshift implementation does not halt the JVM on fatal error
After getting a StackOverflowError I observed that my application (using IOApp) stopped working, but did not halt the JVM. I am running my app in kubernetes, and a JVM exit would result in the pod to be restarted, but since my IO stopped working, but the JVM still was running the result was basically that nothing was done in my app.

After discussing this in your gitter, receiving really good help we where able to boil this down to a simple example demonstrating the behavior, and a possible solution.

In my view I cannot see any good reason for the IOApp to continue to run after a fatal error, so I propose that the default ContextShift should exit the JVM in such situations.

The issue: 
```
object ButWhy extends IOApp {
  override def run(args: List[String]): IO[ExitCode] = {
      val willFail = IO {
        if (true)
          throw new StackOverflowError("This will fail my code, skip error handler, and not exit the JVM")
        else
          throw new RuntimeException("This will fail the code and be handled by the error handler")
        "Success"
      }.handleErrorWith(t => IO("Failed"))

      Stream
        .eval(willFail)
        .to(_.evalMap(d => IO(println(d))))
        .repeat
        .compile
        .drain
        .as(ExitCode.Success)
  }
}
```

workaround:

```
object ButWhy extends IOApp {
  val log: Logger = LoggerFactory.getLogger(ButWhy.getClass)
  val mainEc: ExecutionContextExecutorService = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(2))

  def exitingEC(wrapped: ExecutionContext): ExecutionContext = new ExecutionContext {
    def execute(r: Runnable): Unit =
      wrapped.execute { () =>
        try r.run()
        catch {
          case t: Throwable =>
            log.error("Unhandled error occured.. Shutting down", t)
            System.exit(-1)
        }
      }
    def reportFailure(cause: Throwable): Unit =
      wrapped.reportFailure(cause)
  }

  override implicit protected def contextShift: ContextShift[IO] = IO.contextShift(exitingEC(mainEc))

  override def run(args: List[String]): IO[ExitCode] = {
    val willFail = IO {
      if (true)
        throw new StackOverflowError("This will fail my code, skip error handler, and not exit the JVM")
      else
        throw new RuntimeException("This will fail the code and be handled by the error handler")
      "Success"
    }.handleErrorWith(t => IO("Failed"))

    Stream
      .eval(willFail)
      .to(_.evalMap(d => IO(println(d))))
      .repeat
      .compile
      .drain
      .as(ExitCode.Success)
  }
}
```

### PR DESCRIPTION
**Pull request by djspiewak**
This pulls `IOApp` off the global `ExecutionContext` and onto a custom one that has better properties. Specifically:

- Crashes on truly fatal errors (rather than doing something horrible, like swallowing `OutOfMemoryError`)
- Lower-bounds the pool to 2 threads (global can go down to 1 if you only have 1 core). Having a 1-thread pool can cause deadlocks in poorly written code. Poorly written code is poor, but I'd rather not give them more headaches, and it's enough of an edge case that I'm okay with it.
- Avoids issues with global pool pollution by third-party code
- Avoids fork f***ing join, which is horrible for this kind of stuff

Related to #347. I'd like to hear from @rossabaker (and ideally @alexandru) as to their thoughts on the design tradeoffs here, and whether I should expand the impact at all.

I want to get this in 2.0 because, while it isn't breaking in the traditional sense, it's most definitely a very subtle compatibility shift and I don't want to catch people off guard.

Fixes #509 
Fixes #515

### CODE REVIEW

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.concurrent.atomic.AtomicInteger
+
+private[internals] object PoolUtils {
+
+  // we can initialize this eagerly because the enclosing object is lazy
+  val ioAppGlobal: ExecutionContext = {
+    // lower-bound of 2 to prevent pathological deadlocks on virtual machines
+    val bound = math.max(2, Runtime.getRuntime().availableProcessors())
+
+    val executor = Executors.newFixedThreadPool(bound, new ThreadFactory {
```
Comment (rossabaker) [inline core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Is the basic argument against ForkJoin that most people aren't doing divide-and-conquer tasks, and therefore we're all paying for the overhead of the layered queues without reaping its rewards?  I intuitively agree, but Scala standard library chose otherwise, and the [`ForkJoinPool` javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html) claims:

> efficient processing ... when many small tasks are submitted to the pool from external clients

and

> Especially when setting asyncMode to true in constructors, ForkJoinPools may also be appropriate for use with event-style tasks that are never joined.

In _theory_, I agree with this change.  Empirically, I've swapped them out on real world workloads and not seen much difference either way.  I'm curious if you have apps where a FJ pool is noticeably detrimental to performance.  (Other points, like catching fatals and a minimum of one, are well taken and discussed elsewhere.  I'm strictly speaking performance here.)

I imagine most cats-effect users implement blocking with `evalOn` and a second pool, but those who don't and whose blocking libraries save them by assuming a `blocking`-integrated executor will be in for an unpleasant surprise.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      val ctr = new AtomicInteger(0)
+      def newThread(r: Runnable): Thread = {
+        val back = new Thread(r)
+        back.setName(s"ioapp-compute-${ctr.getAndIncrement()}")
+        back.setDaemon(true)
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: I went back and forth on this. I think it's okay for these to be daemon specifically *because* we have other control flow mechanisms which guarantee everything gets wound down before exiting under normal circumstances. In the event that something is exiting without tracing through the control flow graph all the way to the end of `IOApp`, then something horrible has happened (like OOM), and we want the threads to just die.

TLDR, I think that threads blocking shutdown is a feature that isn't relevant or necessary in applications constructed with explicit monadic control flow.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          try {
+            r.run()
+          } catch {
+            case NonFatal(t) =>
+              reportFailure(t)
+
+            case t: Throwable =>
+              // under most circumstances, this will work even with fatal errors
+              t.printStackTrace()
+              System.exit(-1)
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Will change.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  def exitOnFatal(ec: ExecutionContext): ExecutionContext = new ExecutionContext {
+    def execute(r: Runnable): Unit = {
+      ec.execute(new Runnable {
+        def run(): Unit = {
+          try {
+            r.run()
+          } catch {
+            case NonFatal(t) =>
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: That's a fair point. Also that's a good example of an OOM which is legit recoverable. Another false-positive, as a note, could be `AssertionError`. It's another problem that I went back and forth on.

I think the thing is that, the *moment* some sort of truly fatal error happens, I really do want to know it and explode spectacularly. The worst feeling in the world is knowing that your system *probably* went OOM (or had a link error, or such), but it was swallowed and now your only way to figure it out is to dump the running system and look for orphans.

File: core/jvm/src/main/scala/cats/effect/internals/IOAppPlatform.scala [hunk 5]
```diff
@@ -48,8 +48,11 @@ private[effect] object IOAppPlatform {
 
   // both lazily initiated on JVM platform to prevent
   // warm-up of underlying default EC's for code that does not require concurrency
-  def defaultTimer: Timer[IO] = IOTimer.global
-  def defaultContextShift: ContextShift[IO] = IOContextShift.global
+  def defaultTimer: Timer[IO] =
+    IOTimer(PoolUtils.ioAppGlobal)
```
[the remark this answers is not in the document]
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/IOAppPlatform.scala]: I wasn't entirely clear on the implementation, but it looks like it's already doing that.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala (hunk 1 above)
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: > I'm curious if you have apps where a FJ pool is noticeably detrimental to performance.

Very yes. There isn't a good tldr response to this comment, so I'll write something up in a gist instead.

Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Oh actually, #509 contains a halfway-decent description. I can flesh it out more if you like. It's a lot less about the layered queues and more about the spawning semantics being optimized for bad thread pool usage. ForkJoin is a fantastic *general purpose* pool for Java. It's not a terrible choice for `Future`, either, since people tend to abuse `Future` for blocking side effects and such. Users of cats-effect though tend to be a lot more scrupulous about their pools, since the API is very carefully (and very effectively) designed to carrot them to do so.

When you use ForkJoin to exclusively perform CPU-bound tasks that are throughput-optimized, it will behave *horrifically* poorly and you'll end up with a zillion threads.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (rossabaker) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: And I suppose we can say that any service that relies on its thread pools' graceful shutdown needs to be run on a separate, non-daemon pool.

File: core/jvm/src/main/scala/cats/effect/internals/IOAppPlatform.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (rossabaker) [reply core/jvm/src/main/scala/cats/effect/internals/IOAppPlatform.scala]: Oh, right, because the `ScheduledExecutorService` is a second argument in an overload, and the unary version calls it with a lazy `scheduler.`  Never mind.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala (hunk 1 above)
Comment (rossabaker) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Do you tend to see this with longer-running CPU intensive tasks, rather than a many small ones?

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Exactly.

File: core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala (hunk 1 above)
Comment (djspiewak) [reply core/jvm/src/main/scala/cats/effect/internals/PoolUtils.scala]: Very much so, yes. Many small ones are generally okay because you're yielding back to the pool frequently. Heavily fairness-optimized workflows also don't see this issue. I'm relatively certain, based on the symptoms, that fork-join's heuristics are simply mistakenly identifying throughput-optimized workflows as being blocking IO, and thus spinning up new threads to avoid starvation.
